core/trino-main/src/test/java/io/trino/operator/scalar/TestConditions.java

```diff
@@ -157,12 +156,11 @@ public void testIn()
         assertFunction("(3 in (2, null)) is null", BOOLEAN, true);
         assertFunction("(null not in (2, null, 3, 5)) is null", BOOLEAN, true);
         assertFunction("(3 not in (2, null)) is null", BOOLEAN, true);
-    }
 
-    @Test
-    public void testInDoesNotShortCircuit()
-    {
-        assertInvalidFunction("3 in (2, 4, 3, 5 / 0)", DIVISION_BY_ZERO);
+        // Short circuit: this expression is optimized by SimplifyExpressions to 'true', so it does not fail due to division by zero.
+        // However, the test fails, because one of it parts is "interpret", which tries to evaluate the unoptimized expression
+        // and throws DIVISION_BY_ZERO.
+        /*assertFunction("3 in (2, 4, 3, 5 / 0)", BOOLEAN, true);*/
```

Comment:
`SimplifyExpressions` optimizes it to `true`.
This is done by `ExpressionInterpreter.optimize()`. It indeed evaluates all expressions in the IN-list, but error is suppressed, and eventually the match is found.
Is it incorrect that it succeeds?

Reply:
Yes, it's incorrect because the order of evaluation should consider first all the elements in the list and only *then* do the containment check.

Reply:
Okay, so we've got a bug. According to what you said, this query should fail.
```
trino:tiny> SELECT * FROM region WHERE 1 IN (1, 2, regionkey / regionkey);
 regionkey |    name     |                                                       comment
-----------+-------------+---------------------------------------------------------------------------------------------------------------------
         0 | AFRICA      | lar deposits. blithely final packages cajole. regular waters are final requests. regular accounts are according to
         1 | AMERICA     | hs use ironic, even requests. s
         2 | ASIA        | ges. thinly even pinto beans ca
         3 | EUROPE      | ly final courts cajole furiously final excuse
         4 | MIDDLE EAST | uickly special accounts cajole carefully blithely close requests. carefully final asymptotes haggle furiousl
(5 rows)
```

Reply:
I discovered that in case when some in-list expression causes exception, we currently:
- either fail the query in `SimplifyExpression` which invokes `ExpressionInterpreter` (that covered "0/0")
- or the query fails in execution
- or the query does not fail at all.

My change took away the first possibility.
I hoped that if the "failing" in-predicate survives until execution, it will fail there. But that was too optimistic. The case `3 in (2, 4, 3, 5 / 0)` slipped through. 

I looked into `InCodeGenerator`, which is responsible for the execution of in-predicates. It is quite complicated. It is optimized for different lengths of the in-list and for different types. **One of the options, named `DIRECT_SWITCH`, drops the non-constant elements to "default bucket" and evaluates them lazily**.
According to what you said, it is a bug. I think that we should investigate `InCodeGenerator`and make sure that he always evaluates all values.

Reply:
What should I do regarding this PR? 
It's no good until `InCodeGenerator` is fixed, because it exposes the issue to a great extent.
Or maybe I should add some "sanity checker" to fail in-predicates before execution?